server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java

```diff
@@ -0,0 +1,343 @@
+    if (!component.isPresent()) {
+      throw new RowNotFoundException(String.format("Component key '%s' not found", key));
+    }
+    return component.get();
+  }
+
+  public Optional<ComponentDto> selectByKey(DbSession session, String key) {
+    return Optional.fromNullable(mapper(session).selectByKey(key));
+  }
+
+  public List<UuidWithProjectUuidDto> selectAllViewsAndSubViews(DbSession session) {
+    return mapper(session).selectUuidsForQualifiers(Qualifiers.VIEW, Qualifiers.SUBVIEW);
+  }
+
+  public List<String> selectProjectsFromView(DbSession session, String viewUuid, String projectViewUuid) {
+    return mapper(session).selectProjectsFromView("%." + viewUuid + ".%", projectViewUuid);
+  }
+
+  /**
+   * Returns all projects (Scope {@link org.sonar.api.resources.Scopes#PROJECT} and qualifier
+   * {@link org.sonar.api.resources.Qualifiers#PROJECT}) which are enabled.
+   *
+   * Used by Views.
+   */
+  public List<ComponentDto> selectProjects(DbSession session) {
+    return mapper(session).selectProjects();
+  }
+
+  /**
+   * Select all root components (projects and views), including disabled ones, for a given organization.
+   */
+  public List<ComponentDto> selectAllRootsByOrganization(DbSession dbSession, String organizationUuid) {
+    return mapper(dbSession).selectAllRootsByOrganization(organizationUuid);
+  }
+
+  /**
+   * Select a page of provisioned (root) components. Results are ordered by ascending name.
+   * @param dbSession
+   * @param organizationUuid uuid of the organization
+   * @param textQuery optional text query to match component name or key
+   * @param qualifiers filter on qualifiers. Must not be null nor empty
+   * @param rowBounds pagination
+   */
+  public List<ComponentDto> selectProvisioned(DbSession dbSession, String organizationUuid, @Nullable String textQuery, Set<String> qualifiers, RowBounds rowBounds) {
+    checkArgument(!qualifiers.isEmpty(), "qualifiers must not be empty");
+    return mapper(dbSession).selectProvisioned(organizationUuid, buildUpperLikeSql(textQuery), qualifiers, rowBounds);
+  }
+
+  /**
+   * Count number of provisioned (root) components.
+   * @param dbSession
+   * @param organizationUuid uuid of the organization
+   * @param textQuery optional text query to match component name or key
+   * @param qualifiers filter on qualifiers. Must not be null nor empty
+   */
+  public int countProvisioned(DbSession dbSession, String organizationUuid, @Nullable String textQuery, Set<String> qualifiers) {
+    checkArgument(!qualifiers.isEmpty(), "qualifiers must not be empty");
+    return mapper(dbSession).countProvisioned(organizationUuid, buildUpperLikeSql(textQuery), qualifiers);
+  }
+
+  @CheckForNull
+  private static String buildUpperLikeSql(@Nullable String textQuery) {
+    if (isBlank(textQuery)) {
+      return null;
+    }
+    return buildLikeValue(textQuery.toUpperCase(Locale.ENGLISH), BEFORE_AND_AFTER);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') NullPointerException might be thrown as 'textQuery' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)